### ISSUE TITLE
container variations GUI from other mods

### ISSUE THREAD

**Issue by DadoIrie** [opekope2/OptiGUI#68]
I'm wondering how to show GUI variations for modded containers like [Variant Barrels](https://www.curseforge.com/minecraft/mc-mods/variant-barrels-fabric) & [More Chest Variants](https://www.curseforge.com/minecraft/mc-mods/more-chest-variants-lieonlion)

**Comment by opekope2** [comment]
Press F3+H to show advanced item tooltips.
You see an ID under the item's name (formatted `modid:item_id`, colored blue iirc), it should be placed inside the brackets (`[modid:item_id]`). Additionally, the `interaction.texture` needs to be specific, because OptiGUI only knows vanilla containers' default textures.
An inspector will be added in the 2.2.0 release (or a pre-release), which will tell you these details.

**Comment by DadoIrie** [comment]
not sure that I can completely follow here: "Additionally, the interaction.texture needs to be specific"
Also I am a resourcepack newbie and not a modder, that been said I don't know if those mods need to have a OptiGUI compatibility coded in them or does OptiGUI handle this?

**Comment by DadoIrie** [comment]
![image](https://github.com/opekope2/OptiGUI/assets/61109153/36d03865-acc9-4b9c-8e1a-cb410dca367d)
![image](https://github.com/opekope2/OptiGUI/assets/61109153/6879b03f-8caa-4413-b6a1-a9b44aad4bb3)
What I'm missing here? As said - not getting what you mean with "the interaction.texture needs to be specific"
```
[lolmcv:bamboo_chest]
replacement=./bamboo.png
```

**Comment by DadoIrie** [comment]
what I'm doing wrong here:

tried both folder namespaces
![image](https://github.com/opekope2/OptiGUI/assets/61109153/5149420a-5dc1-4558-bed1-b21a9fe8169a)
![image](https://github.com/opekope2/OptiGUI/assets/61109153/9abf7360-30e7-4773-a8e3-b044e714cdf9)
with
```
[lolmcv:bamboo_chest]
interaction.texture=minecraft:textures/gui/container/generic_54.png
replacement=./bamboo.png
```

**Comment by opekope2** [comment]
Your last snippet looks right, I know why it doesn't work.
A small amount of code is needed to connect the two mods. It can be done on either OptiGUI side or the other side.
The current codebase makes it quite simple to add this. I'll add it to OptiGUI in some way in 2.2.0.
If you find other custom container mod, let me know in another issue, so I can add support for them.

**Comment by opekope2** [comment]
![More Chest Variants test](https://github.com/opekope2/OptiGUI/assets/20487556/b5dde43d-7e2e-4d40-a108-06a3abf6fd7f)
I got MCV working with my modified [resource pack](https://modrinth.com/resourcepack/chest-boat-gui). The code is not yet upstream.

**Comment by DadoIrie** [comment]
yay - can't wait for it 🥇 
snipped is right?
```
[lolmcv:bamboo_chest]
interaction.texture=minecraft:textures/gui/container/generic_54.png
replacement=bamboo_chest.png
```

**Comment by opekope2** [comment]
Yep, except `interaction.texture` won't be required

**Comment by DadoIrie** [comment]
and each mod which adds containers needs additional compatibility or is that also taken care of?

**Comment by opekope2** [comment]
For simple mods (like MCV), only a couple lines of code is needed, which processes the block entity (OptiGUI can only process vanilla block entities by default).
For completely new container, which add other properties (for example, a color or variant), it will need a selector (like `myblock.color`) and an interface+implementation containing it implemented in OptiGUI. I'll work on selector extensibility.
This code can be added to either the other mod or OptiGUI (open a PR if you find such a mod, and I'll probably make it compatible... eventually).
I can't reference any code or docs because I haven't pushed them yet.

**Comment by DadoIrie** [comment]
then for now please add compatibility for MCV and the Variant Barrels ... or [Variant Vanilla Blocks](https://legacy.curseforge.com/minecraft/mc-mods/variant-vanilla-blocks) which has basically the crafting tables, grindstones, barrels, cartography tables, smithing tables & smokers ... actually (also don't wanna come over like requesting features and compatibilities) since I'm no modder and have some basic knowledge of Javascript (meaning I can read and modify code - I've also once successfully extracted a mods single functionality which also worked as a standalone mini mod) and if I get some examples (meaning if it's one file where I just need to rename some references) I'd look into it adding the [Variant Vanilla Blocks](https://legacy.curseforge.com/minecraft/mc-mods/variant-vanilla-blocks) which are using the vanilla GUI of the corresponded container.

**Comment by opekope2** [comment]
I still have to find a good way to automatically test core functionality, and I have many ideas to implement in the next release. Compatibly is just one thing I'm working on. I'll eventually get there.

**Comment by opekope2** [comment]
I just took a look at variant barrels
There are more variant mods to support
Will take some time
(also blaming the slow development pace on gradle)

**Comment by opekope2** [comment]
Whoops
That's why I don't usually push individual commits

**Comment by DadoIrie** [comment]
why?

**Comment by opekope2** [comment]
Because I can fix things if I miss something and I don't need to add a new commit or force push

**Comment by DadoIrie** [comment]
from the creator of Variant Barrels & Variant Vanilla Block: [Variant Chests](https://www.curseforge.com/minecraft/mc-mods/variantchests)

**Comment by FernLeafie** [comment]
hello, i just stumbled across this while trying to do something similar for the [farmer's delight](https://modrinth.com/mod/farmers-delight-fabric) cabinets, giving them each a different texture depending on the wood type, and if i understand this correctly this wouldn't be possible at the current time. so could i suggest that that could be looked into in the future?

**Comment by opekope2** [comment]
I'll open a PR in it, when I release OptiGUI 2.2.0. The first alpha build contains the API, which may be refined later on.
It's been quiet here, because I went on to implement the Toast abstraction into LiLaC, then I did university things, and now I develop Staff Mod. If I'm lucky, I'll get this done by Minecraft 1.21, but no promises. Yes, I'm still alive.

**Comment by Aulberon** [comment]
I'm trying to add support for quark's variant chests but it doesn't work.
This is how I'm doing:

```
[quark:acacia_chest]
interaction.texture = minecraft:textures/gui/container/generic_54.png
replacement = assets/quark/acacia_chest.png

[chest quark:ancient_chest]
interaction.texture = minecraft:textures/gui/container/generic_54.png
replacement = assets/quark/ancient_chest.png
```

None works
What am I doing wrong?

**Comment by opekope2** [comment]
If the INI file is at `assets/optigui/gui/whatever.ini`, OptiGUI looks for the replacement texture at `assets/optigui/gui/assets/quark/acacia_chest.png`, because you specified a [relative path](https://opekope2.dev/OptiGUI/syntax.html#paths). If the replacement texture is at `assets/quark/acacia_chest.png`, `quark:acacia_chest.png` should be specified as replacement.

**Comment by Aulberon** [comment]
Wait, so the replacement is the target texture that I want to change?

The texture I made to replace quark's is at "assets/optigui/gui/assets/quark/acacia_chest.png".

Can you give me an example how to properly do it?

**Comment by opekope2** [comment]
Nope, OptiGUI looks for the replacement in the loaded resource packs. I assumed you placed the resources in the wrong folder, so your example looks correct, if every file is in the right place.
If OptiGUI can't find the `replacement` while loading resource packs, it will log a warning in the game's console to let you know:
![warning](https://github.com/opekope2/OptiGUI/assets/20487556/04e23295-37d1-4e75-b397-323a5eda96ab)
I will eventually add an in-game GUI for this.

`interaction.texture` tells OptiGUI what the original texture is. This is optional for vanilla containers, because OptiGUI knows their original texture.
`replacement` tells OptiGUI what to replace it with.
For that texture, the absolute path (ID) is `optigui:gui/assets/quark/acacia_chest.png`, or you can use a relative path starting at the INI file's folder.

The support for other mods is very basic and not very well-tested, so it could be an OptiGUI issue, but it got better in 2.3.0-alpha.1.

**Comment by Aulberon** [comment]
The problem seems to be with the target.
The "[chest quark:ancient_chest]" actually changed the vanilla chest texture, then I tested some textures with the vanilla chest and everything worked, the problem seems to be the "[quark:acacia_chest]".
Looking at the "latest.log" the only reference to optigui load failure is with the "desktop.ini".

Maybe creating a single .ini file for each chest works? Right now I'm using a single .ini to change a bunch of things (vanilla chests, villagers, shulkers and more).

**Comment by opekope2** [comment]
Are you running OptiGUI 2.1.x on Sinytra?

**Comment by Aulberon** [comment]
Yes

**Comment by opekope2** [comment]
Then ignore what I said, it's about OptiGUI 2.3.0.
OptiGUI 2.1.x is an ancient codebase, it requires a preprocessor to be registered for each block entity class. This preprocessor was removed in 2.3.0-alpha.1 for simplicity and better compatibility.
Sinytra is not tested at all.

**Comment by Aulberon** [comment]
Is 2.3.0-alpha.1 compatible with 1.20.1?
I could test it!

**Comment by opekope2** [comment]
Not yet, but I will backport to 1.20.1 after the full release (because OptiGUI 2.3.0+ will not be a universal JAR to support 1.18-1.21)

**Comment by Aulberon** [comment]
I will try and test some things on newer versions, so when the backport drops, I can compare and test.

**Comment by opekope2** [comment]
The first alpha builds of OptiGUI 3 are finally here for 1.20 and 1.21!

OptiGUI 3 uses NBT instead of selectors, which is way more compatible with other mods out-of-the-box. Press `F12` while the inventory is open, and OptiGUI will generate and copy a [JSON Resource](https://opekope2.dev/OptiGUI/JSON.html) to your clipboard.
